File: Examples/Analysis_4/Analysis_4.py
# -*- coding: utf-8 -*-
"""
Multivariate Bayesian Inversion for Classification and Regression
Analysis 4: brain age prediction (Python script)

Author: Joram Soch, OvGU Magdeburg
E-Mail: joram.soch@ovgu.de

Version History:
- 21/02/2022, 20:29: first version in MATLAB
- 27/02/2025, 18:01: ported code to Python (Step 1 & 2)
- 28/02/2025, 11:16: ported code to Python (Step 3)
- 05/03/2025, 11:04: added to GitHub repository
- 05/03/2025, 18:45: unified tick font sizes
"""


# specify MBI path
MBI_dir = '../../Python/'

# import packages
import os
import numpy as np
import scipy as sp
from sklearn import svm
import matplotlib.pyplot as plt

# import MBI module
orig_dir = os.getcwd()
os.chdir(MBI_dir)
import MBI
os.chdir(orig_dir)


### Step 1: load data #########################################################

# load data
train = sp.io.loadmat('PAC_specify.mat')
test  = sp.io.loadmat('PAC_specify_test_age.mat')

# assemble data (MBR)
n1  = train['sID1'].size                    # number of data points
n2  = train['sID2'].size
V1  = np.eye(n1)                            # observation covariances
V2  = np.eye(n2)
YA1 = np.c_[train['GM1'], train['WM1']]     # data matrices
YA2 = np.c_[train['GM2'], train['WM2']]
x1  = np.squeeze(train['y1'])               # label vectors
x2  = np.squeeze(test['y2'])
XA1 = train['c1'][:,1:]                     # covariate matrices
XA2 = train['c2'][:,1:]

# assemble data (SVR)                       # feature matrices
YB1 = np.c_[train['GM1'], train['WM1'], train['c1']]
YB2 = np.c_[train['GM2'], train['WM2'], train['c2']]
del train, test

# prepare histograms
x_min = 0
x_max = 100
dx    = 2.5
nx    = int((x_max-x_min)/dx)+1
xb    = np.linspace(x_min, x_max, nx)


### Step 2: analyze data ######################################################

# estimate gamma distribution
x1_min = np.min(x1)-0.5
ab_est = sp.stats.gamma.fit(x1-x1_min, floc=0)

# compute training histogram
xp = np.arange(x_min, x_max+1, 1)
fb = np.concatenate((xp-0.5, np.array([xp[-1]+0.5])))
f1 = np.histogram(x1, bins=fb)[0]
del fb

# define priors for MBR
priors = ['uniform', 'data-driven', 'fitted']
prior  = [{} for pr in priors]
prior[0] = {'x': xp, 'p': (1/np.ptp(xp)) * np.ones(xp.shape)}
prior[1] = {'x': xp, 'p': (1/np.trapz(f1, xp)) * f1}
prior[2] = {'x': xp, 'p': sp.stats.gamma.pdf(xp-x1_min, ab_est[0], loc=0, scale=ab_est[2])}

# Analysis 1: MBR with site/sex as covariates
print('\n-> MBR: train, ', end='')
MBA1 = MBI.model(YA1, x1, X=XA1, V=V1, mb_type='MBR').train()
PP2  = [None for pr in priors]
xMAP = np.zeros((n2,len(priors)))
rA   = np.zeros(len(priors))
maeA = np.zeros(len(priors))
print('test: prior ', end='')
for h in range(len(prior)):
    print('{}, '.format(h+1), end='')
    PP2[h]  = MBI.model(YA2, x2, X=XA2, V=V2, mb_type='MBR').test(MBA1, prior[h])
    for i in range(n2):
        xMAP[i,h] = prior[h]['x'][np.argmax(PP2[h][i,:])]
    rA[h]   = np.corrcoef(xMAP[:,h], x2)[0,1]
    maeA[h] = np.mean(np.abs(xMAP[:,h]-x2))
print('done.')

# Analysis 2: SVR with site/sex as features
print('\n-> SVR: train, ', end='')
SVM1 = svm.SVR(kernel='linear', C=1)
SVM1.fit(YB1, x1)
print('test, ', end='')
xp2  = SVM1.predict(YB2)
rB   = np.corrcoef(xp2, x2)[0,1]
maeB = np.mean(np.abs(xp2-x2))
print('done.\n')

# calculate histograms
nb1 = np.histogram(x1, bins=xb)[0]
nb2 = np.histogram(x2, bins=xb)[0]
nbA = np.zeros((len(priors), xb.size-1))
nbB = np.histogram(xp2, bins=xb)[0]
for h in range(len(prior)):
    nbA[h,:] = np.histogram(xMAP[:,h], bins=xb)[0]


### Step 3: visualize results #################################################

# open figure
fig1 = plt.figure(figsize=(16,10))
axs  = fig1.subplots(3,5)

# 1st row
axs[0,0].bar(xb[:-1]+dx/2, nb1,
             width=dx, align='center', color=(3/4,3/4,3/4), edgecolor='k')
axs[0,0].plot(prior[2]['x'], prior[2]['p']*n1*dx, '-k', linewidth=2)
axs[0,0].axis([x_min, x_max, 0, (11/10)*np.max(nb1)])
axs[0,0].set_xlabel('chronological age [yrs]', fontsize=16)
axs[0,0].set_ylabel('number of subjects', fontsize=16)
axs[0,0].set_title('Training Set', fontsize=16, fontweight='bold')
axs[0,0].tick_params(axis='both', labelsize=10)

for h in range(len(prior)):
    axs[0,1+h].plot([x_min, x_max], [x_min, x_max], '-k', linewidth=1)
    axs[0,1+h].plot(x2, xMAP[:,h], '.b', markersize=5)
    axs[0,1+h].axis([x_min, x_max, x_min, x_max])
    axs[0,1+h].set_aspect('equal', adjustable='box')
    axs[0,1+h].set_xlabel('actual age', fontsize=16)
    axs[0,1+h].set_ylabel('predicted age', fontsize=16)
    if h == 1:
        axs[0,1+h].set_title('MBR with site/gender as covariates',
                             fontsize=16, fontweight='bold')
    axs[0,1+h].tick_params(axis='both', labelsize=10)
    axs[0,1+h].text(x_min+5, x_max-5, 'r = {:.2f}, MAE = {:.2f}'.format(rA[h], maeA[h]),
                    fontsize=12, ha='left', va='center')

axs[0,4].plot([x_min, x_max], [x_min, x_max], '-k', linewidth=1)
axs[0,4].plot(x2, xp2, '.r', markersize=5)
axs[0,4].axis([x_min, x_max, x_min, x_max])
axs[0,4].set_aspect('equal', adjustable='box')
axs[0,4].set_xlabel('actual age', fontsize=16)
axs[0,4].set_ylabel('predicted age', fontsize=16)
axs[0,4].set_title('SVR with site/gender as features',
                   fontsize=16, fontweight='bold')
axs[0,4].tick_params(axis='both', labelsize=10)
axs[0,4].text(x_min+5, x_max-5, 'r = {:.2f}, MAE = {:.2f}'.format(rB, maeB),
              fontsize=12, ha='left', va='center')

# 2nd row
axs[1,0].bar(xb[:-1]+dx/2, nb2,
             width=dx, align='center', color=(3/4,3/4,3/4), edgecolor='k')
axs[1,0].axis([x_min, x_max, 0, (11/10)*np.max(nb2)])
axs[1,0].set_xlabel('chronological age [yrs]', fontsize=16)
axs[1,0].set_ylabel('number of subjects', fontsize=16)
axs[1,0].set_title('Validation Set', fontsize=16, fontweight='bold')
axs[1,0].tick_params(axis='both', labelsize=10)

for h in range(len(prior)):
    axs[1,1+h].bar(xb[:-1]+dx/2, nbA[h,:],
                   width=dx, align='center', color=(0,0,1), edgecolor='k')
    axs[1,1+h].axis([x_min, x_max, 0, (11/10)*np.max(nbA[h,:])])
    axs[1,1+h].set_xlabel('predicted age', fontsize=16)
    axs[1,1+h].set_ylabel('number of subjects', fontsize=16)
    if h == 1:
        axs[1,1+h].set_title('MBR: prediction distribution',
                             fontsize=16, fontweight='bold')
    axs[1,1+h].tick_params(axis='both', labelsize=10)

axs[1,4].bar(xb[:-1]+dx/2, nbB,
             width=dx, align='center', color=(1,0,0), edgecolor='k')
axs[1,4].axis([x_min, x_max, 0, (11/10)*np.max(nbB)])
axs[1,4].set_xlabel('predicted age', fontsize=16)
axs[1,4].set_ylabel('number of subjects', fontsize=16)
axs[1,4].set_title('SVR: prediction distribution',
                   fontsize=16, fontweight='bold')
axs[1,4].tick_params(axis='both', labelsize=10)

# 3rd row
for h in range(len(prior)):
    axs[2,1+h].plot(prior[h]['x'], prior[h]['p'], '-b', linewidth=1)
    axs[2,1+h].axis([x_min, x_max, 0, (11/10)*np.max(prior[h]['p'])])
    if h == 0: axs[2,1+h].set_ylim([0, 2*np.max(prior[h]['p'])])
    axs[2,1+h].set_xlabel('chronological age [yrs]', fontsize=16)
    axs[2,1+h].set_ylabel('prior density', fontsize=16)
    axs[2,1+h].set_title('{} prior'.format(priors[h]),
                         fontsize=16, fontweight='bold')
    axs[2,1+h].tick_params(axis='both', labelsize=10)
axs[2,0].axis('off')
axs[2,4].axis('off')

# open figure
fig2 = plt.figure(figsize=(16,10))
axs  = fig2.subplots(3,5)

# all rows
for h in range(len(prior)):
    for i in range(5):
        axs[h,i].plot(x2[i], (1/10)*np.max(PP2[h][i,:]), 'xk',
                      markersize=7.5, linewidth=3, label='true')
        axs[h,i].plot(xMAP[i,h], (1/10)*np.max(PP2[h][i,:]), '.b',
                      markersize=15, label='MAP')
        axs[h,i].plot(prior[h]['x'], PP2[h][i,:], '-b',
                      linewidth=1)
        axs[h,i].axis([np.min(prior[h]['x']), np.max(prior[h]['x']),
                       0, (11/10)*np.max(PP2[h][i,:])])
        if h == 0 and i == 0:
            axs[h,i].legend(loc='upper right')
        if h == 2:
            axs[h,i].set_xlabel('chronological age [yrs]', fontsize=16)
        if i == 1:
            axs[h,i].set_ylabel('posterior density', fontsize=16)
        if i == 0:
            axs[h,i].set_ylabel('{} prior'.format(priors[h]),
                                fontsize=20, fontweight='bold')
        if h == 0:
            axs[h,i].set_title('Subject {}'.format(i+1),
                               fontsize=20, fontweight='bold')
        axs[h,i].tick_params(axis='both', labelsize=10)

# enable tight layout
fig1.tight_layout()
fig2.tight_layout()

File: Examples/Analysis_4/Analysis_4.m
% Multivariate Bayesian Inversion for Classification and Regression
% Analysis 4: brain age prediction (MATLAB script)
% 
% Author: Joram Soch, OvGU Magdeburg
% E-Mail: joram.soch@ovgu.de
% 
% Version History:
% - 21/02/2022, 01:45: data analysis
% - 21/02/2022, 20:29: results visualization
% - 30/05/2022, 05:50: minor changes & finalization
% - 28/02/2025, 18:06: aligned with Python
% - 11/02/2026, 17:19: recorded analysis time


clear
close all

%%% Step 1: load data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% load data
load PAC_specify.mat
load PAC_specify_test_age.mat

% assemble data (MBR)
n1  = numel(sID1);              % number of data points
n2  = numel(sID2);
V1  = eye(n1);                  % observation covariances
V2  = eye(n2);
YA1 = [GM1, WM1];               % data matrices
YA2 = [GM2, WM2];
x1  = y1;                       % label vectors
x2  = y2;
XA1 = c1(:,2:end);              % covariate matrices
XA2 = c2(:,2:end);

% assemble data (SVR)
YB1 = [GM1, WM1, c1];           % feature matrices
YB2 = [GM2, WM2, c2];

% prepare histograms
x_min = 0;
x_max = 100;
dx    = 2.5;
xb    = [(x_min+dx/2):dx:(x_max-dx/2)];


%%% Step 2: analyze data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% estimate gamma distribution
y1_min = min(y1)-0.5;
ab_est = gamfit(y1-y1_min);

% define priors for MBR
priors = {'uniform', 'data-driven', 'fitted'};
prior(1).x = [0:1:100];
prior(1).p = (1/range(prior.x))*ones(size(prior.x));
prior(2).x = [0:1:100];
prior(2).p = hist(y1, prior(2).x)./trapz(prior(2).x, hist(y1, prior(2).x));
prior(3).x = [0:1:100];
prior(3).p = gampdf(prior(3).x-y1_min, ab_est(1), ab_est(2));

% Analysis 1: MBR with site/sex as covariates
tic;
fprintf('\n-> MBR: train, ');
MBA1 = mbitrain(YA1, x1, XA1, V1, 'MBR');
tA   = toc*ones(1,numel(prior));
PP2  = cell(1,numel(prior));
xMAP = zeros(n2,numel(prior));
rA   = zeros(1,numel(prior));
maeA = zeros(1,numel(prior));
fprintf('test: prior ');
for h = 1:numel(prior)
    tic;
    fprintf('%d, ', h);
    PP2{h} = mbitest(YA2, x2, XA2, V2, MBA1, prior(h));
    tA(h)  = tA(h) + toc;
    for i = 1:n2
        xMAP(i,h) = prior(h).x(PP2{h}(i,:)==max(PP2{h}(i,:)));
    end;
    rA(h)   = corr(xMAP(:,h), x2);
    maeA(h) = mean(abs(xMAP(:,h)-x2));
end;
fprintf('done.\n');

% Analysis 2: SVR with site/sex as features
tic;
fprintf('-> SVR: train, ');
SVM1 = svmtrain(x1, YB1, '-s 4 -t 0 -c 1 -q');
fprintf('test, ');
xp2  = svmpredict(x2, YB2, SVM1, '-q');
fprintf('done.\n\n');
tB   = toc;
rB   = corr(xp2, x2);
maeB = mean(abs(xp2-x2));

% calculate histograms
nb1 = hist(y1, xb);
nb2 = hist(y2, xb);
nbA = zeros(numel(prior), numel(xb));
nbB = hist(xp2,  xb);
for h = 1:numel(prior)
    nbA(h,:) = hist(xMAP(:,h), xb);
end;

% store analysis time
time = {'Analysis 4', 'Figure 11, 2nd/5th',  tA(1), tB;
        'Analysis 4', 'Figure 11, 3rd col.', tA(2), NaN;
        'Analysis 4', 'Figure 11, 4th col.', tA(3), NaN;}; 
save('Analysis_4.mat', 'time');


%%% Step 3: visualize results %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% open figure
figure('Name', 'Analysis 3 (1)', 'Color', [1 1 1], 'Position', [50 50 1600 900]);

% 1st row
subplot(3,5,1); hold on;
bar(xb, nb1, 'FaceColor', [3/4, 3/4, 3/4]);
plot(prior(3).x, prior(3).p*n1*dx, '-k', 'LineWidth', 2);
axis([x_min, x_max, 0, (11/10)*max(nb1)]);
set(gca,'Box','On');
xlabel('chronological age [yrs]', 'FontSize', 12);
ylabel('number of subjects', 'FontSize', 12);
title('Training Set', 'FontSize', 16);

for h = 1:numel(prior)
    subplot(3,5,1+h); hold on;
    plot([x_min, x_max], [x_min, x_max], '-k', 'LineWidth', 1);
    plot(x2, xMAP(:,h), '.b', 'MarkerSize', 10);
    axis([x_min, x_max, x_min, x_max]);
    axis square;
    set(gca,'Box','On');
    xlabel('actual age', 'FontSize', 12);
    ylabel('predicted age', 'FontSize', 12);
    if h == 2, title('MBR with site/gender as covariates', 'FontSize', 16); end;
    text(x_min+5, x_max-5, sprintf('r = %0.2f, MAE = %0.2f', rA(h), maeA(h)), ...
         'HorizontalAlignment', 'Left', 'VerticalAlignment', 'Middle');
end;

subplot(3,5,5); hold on;
plot([x_min, x_max], [x_min, x_max], '-k', 'LineWidth', 1);
plot(x2, xp2, '.r', 'MarkerSize', 10);
axis([x_min, x_max, x_min, x_max]);
axis square;
set(gca,'Box','On');
xlabel('actual age', 'FontSize', 12);
ylabel('predicted age', 'FontSize', 12);
title('SVR with site/gender as features', 'FontSize', 16);
text(x_min+5, x_max-5, sprintf('r = %0.2f, MAE = %0.2f', rB, maeB), ...
     'HorizontalAlignment', 'Left', 'VerticalAlignment', 'Middle');

% 2nd row
subplot(3,5,6);
bar(xb, nb2, 'FaceColor', [3/4, 3/4, 3/4]);
axis([x_min, x_max, 0, (11/10)*max(nb2)]);
xlabel('chronological age [yrs]', 'FontSize', 12);
ylabel('number of subjects', 'FontSize', 12);
title('Validation Set', 'FontSize', 16);

for h = 1:numel(prior)
    subplot(3,5,6+h);
    bar(xb, nbA(h,:), 'b');
    axis([x_min, x_max, 0, (11/10)*max(nbA(h,:))]);
    xlabel('predicted age', 'FontSize', 12);
    ylabel('number of subjects', 'FontSize', 12);
    if h == 2, title('MBR: prediction distribution', 'FontSize', 16); end;
end;

subplot(3,5,10);
bar(xb, nbB, 'r');
axis([x_min, x_max, 0, (11/10)*max(nbB)]);
xlabel('predicted age', 'FontSize', 12);
ylabel('number of subjects', 'FontSize', 12);
title('SVR: prediction distribution', 'FontSize', 16);

% 3rd row
for h = 1:numel(prior)
    subplot(3,5,11+h);
    plot(prior(h).x, prior(h).p, '-b', 'LineWidth', 1);
    axis([x_min, x_max, 0, (11/10)*max(prior(h).p)]);
    if h == 1, ylim([0, 2*max(prior(h).p)]); end;
    xlabel('chronological age [yrs]', 'FontSize', 12);
    ylabel('prior density', 'FontSize', 12);
    title(sprintf('%s prior', priors{h}), 'FontSize', 16);
end;

% open figure
figure('Name', 'Analysis 3 (2)', 'Color', [1 1 1], 'Position', [50 50 1600 900]);

% all rows
for h = 1:numel(prior)
    for i = 1:5
        subplot(3,5,(h-1)*5+i); hold on;
        plot(x2(i), (1/10)*max(PP2{h}(i,:)), 'xk', 'MarkerSize', 7.5, 'LineWidth', 2);
        plot(xMAP(i,h), (1/10)*max(PP2{h}(i,:)), '.b', 'MarkerSize', 20);
        plot(prior(h).x, PP2{h}(i,:), '-b', 'LineWidth', 1);
        axis([min(prior(h).x), max(prior(h).x), 0, (11/10)*max(PP2{h}(i,:))]);
        set(gca,'Box','On');
        if h == 1 && i == 1, legend('true', 'mode', 'Location', 'NorthEast'); end;
        if h == 3, xlabel('chronological age [yrs]', 'FontSize', 12); end;
        if i == 1, ylabel(sprintf('%s prior', priors{h}), 'FontSize', 16, 'FontWeight', 'Bold'); end;
        if i == 2, ylabel('posterior density', 'FontSize', 12); end;
        if h == 1, title(sprintf('Subject %d', i), 'FontSize', 16); end;
    end;
end;